### PR TITLE
fix issue 57

### PR DESCRIPTION
**Pull request by hatufacci**
fix issue 57, handle offset test rail api limitation

### CODE REVIEW

File: lib/testrail.js [hunk 1]
```diff
@@ -21,10 +21,31 @@ class TestRail {
...
+		let fullArr = [];
 		try {
 			const res = await this.axios.get(`get_cases/${projectId}&suite_id=${suiteId}`);
 			output.log(`getCases: SUCCESS - the response data is ${JSON.stringify(res.data)}`);
-			return res.data.cases;
+			let isLimit = false
+			let offset = 250;
+			fullArr = res.data.cases;
+			//limit logic works only if we suspect that api call send back limited array === 250 cases
+			if(res.data.cases.length === offset){
```
Comment (vladbalanescu) [inline lib/testrail.js:32]: no, this is bad. We cannot simply relie on this logic to tell if the user is on a specific TR version or not. For example, what if I'm on `7.0.1` which is what I'm currently on and I have strictly 250 cases. That'd break for me, so we need another way to figure this one out.

File: lib/testrail.js [hunk 2]
```diff
@@ -21,10 +21,31 @@ class TestRail {
...
 
 	async getCases(projectId, suiteId) {
+		let fullArr = [];
 		try {
 			const res = await this.axios.get(`get_cases/${projectId}&suite_id=${suiteId}`);
 			output.log(`getCases: SUCCESS - the response data is ${JSON.stringify(res.data)}`);
-			return res.data.cases;
+			let isLimit = false
+			let offset = 250;
+			fullArr = res.data.cases;
```
Comment (vladbalanescu) [inline lib/testrail.js:30]: `fullArr = res.data && res.data.cases`  <- in case `data` is undefined

**Review by vladbalanescu — CHANGES_REQUESTED** [review]
The rest of the logic seems alright to me, but we need to change that logic in the conditional clause.

**Comment by PeterNgTr** [comment]
@hatufacci Could you find out what version of testrail are you using and then we compare with @vladbalanescu.

**Comment by hatufacci** [comment]
> @hatufacci Could you find out what version of testrail are you using and then we compare with @vladbalanescu.

my version is 6.2.3.114.

**Comment by PeterNgTr** [comment]
Seemed like this is the latest version https://secure.gurock.com/customers/testrail/version/check/7.0.2.1015

**Comment by hatufacci** [comment]
> https://secure.gurock.com/customers/testrail/version/check/7.0.2.1015

**Comment by hatufacci** [comment]
> Seemed like this is the latest version https://secure.gurock.com/customers/testrail/version/check/7.0.2.1015

**Comment by hatufacci** [comment]
@PeterNgTr did u have a chance to try with fresh TR version?

File: lib/testrail.js (hunk 1 above)
Comment (hatufacci) [reply lib/testrail.js:32]: what is better have separated flag, or make just one extra call if cases_id === 250, and don't care about versions at all. It's very rare case and set separated flag for that one case it's overhead. You config right now has a lot of flags and should be more easy to use, for example "closeTestRun" should be true by default. Who need opened (not closed) runs after test failed/passed ? @vladbalanescu @PeterNgTr

**Comment by EgorBodnar** [comment]
@hatufacci, @PeterNgTr 
I've this latest TestRail version and more then 1000 cases. This code won't work, issue is in `validResults` in the `index.js`.

If I'll change 
```js
testrail.addResultsForCases(runId, { results: validResults }).then(res => {
```
to 
```js
testrail.addResultsForCases(runId, { results: allResults }).then(res => {
```
it works and push results to TestRail. 
 Please fix the `validResults` filter.

**Comment by hatufacci** [comment]
> ```js
> testrail.addResultsForCases(runId, { results: validResults })
> ```

@EgorBodnar you suggested just delete checking valid results but  i think check for valid results is need. Yes, if you use last version of test rail all your cases, except 250 not valid - and my fix wold check it and would make them valid.

**Comment by EgorBodnar** [comment]
> ```js
> ```js
> testrail.addResultsForCases(runId, { results: validResults })
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> 
> @EgorBodnar you suggested just delete checking valid results but  i think check for valid results is need. Yes, if you use last version of test rail all your cases, except 250 not valid - and my fix would check it and would make them valid.
> ```

Nope. The check shouldn't be deleted. I just pointed out where the error is and that it needs to be fixed.

**Comment by hatufacci** [comment]
> > ```js
> > ```js
> > testrail.addResultsForCases(runId, { results: validResults })
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @EgorBodnar you suggested just delete checking valid results but  i think check for valid results is need. Yes, if you use last version of test rail all your cases, except 250 not valid - and my fix would check it and would make them valid.
> 
> Nope. The check shouldn't be deleted. I just pointed out where the error is and that it needs to be fixed.

unfortunately, i've stuck with this fix, coz guys have doubts  with this extra call in one rare case

**Comment by EgorBodnar** [comment]
@PeterNgTr, is it really need to `validResults` (#44 )? If yes,  @vladbalanescu suggest changes please according to the new version of the TestRail.  Need to solve the issue ASAP because the plugin has been inoperable for a long time.

**Comment by hatufacci** [comment]
@PeterNgTr any updates?

**Pull request merged**